### PR TITLE
Add IDOBAO ID87.

### PR DESCRIPTION
**Pull request by mehoron**
## Description

Adding the ID87 board support to QMK repository as well as adding VIA support. This is a keyboard sold through IDOBAO and utilizes a YMDK PCB.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/id87/readme.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+Hardware URL: [Idobao](https://www.idobao.net/products/idobao-id87-80-hot-swappable-mechanical-keyboard-kit?variant=33990062604419)
+
+This board can enter DFU mode via FN(The menu button)+ESC
+
+Make example for this keyboard (after setting up your build environment):
+
+    make id87:default
+
+See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
```
Comment (sigprof) [inline keyboards/id87/readme.md:15]: Please update this to match [`quantum/template/avr/readme.md`](https://github.com/qmk/qmk_firmware/blob/master/quantum/template/avr/readme.md) (maybe you used an older version of template).

You could probably drop the `?variant=33990062604419` part from the URL.

**Review by sigprof — CHANGES_REQUESTED** [review]

File: keyboards/id87/id87.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	K010, K011, K012, K013, K014, K015, K016, K017, K018, \
+	K020, K021, K022, K023, K024, K025, K026, K027, K028, \
+	K030, K031, K032, K033, K034, K035, K036, K037, K038, \
+	K040,       K042, K043, K044, K045, K046, K047, K048, \
+	K050, K051, K052, K053, K054,       K056, K057, K058, \
+	      K061, K062, K063, K064, K065, K066, K067, K068, \
+	K070, K071, K072, K073,       K075, K076, K077, K078, \
+	      K081, K082, K083, K084, K085, K086, K087, K088, \
+	            K092, K093,       K095,       K097, K098, \
+	      K101, K102,       K104, K105, K106, K107, K108  \
```
[the remark this answers is not in the document]
Comment (mehoron) [reply keyboards/id87/id87.h:30]: I don't have the json file for it yet, I'll try to get a hold of it. I just received a base firmware. I'll do what I can to fix it up. Thanks for taking a look.

**Comment by sigprof** [comment]
I found the ID87 definition at http://mtkeyboard.vip/: open that page (be patient, the site loads really slow), select the “id87” button, then click “SETTINGS” below the keyboard area, scroll down and click the yellow button with Chinese text below the “Save your layout” label — you will get `id87.json`. Upload that JSON to https://noroadsleft.github.io/kbf_qmk_converter/, and you will get usable id87.h and keymap.c from there (just adjust the macro name to `LAYOUT_tkl_ansi`).

File: keyboards/id87/config.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+#define BACKLIGHT_LEVELS 3
+#endif
+
+/* Set 0 if debouncing isn't needed */
+#define DEBOUNCE 5
+
+#define RGB_DI_PIN E2
+#ifdef RGB_DI_PIN
+#define RGBLIGHT_ANIMATIONS
+#define RGBLED_NUM 30
```
[the remark this answers is not in the document]
Comment (mehoron) [reply keyboards/id87/config.h]: The number of LEDs is 22 16 on the bottom and 6 on the top.

**Comment by mehoron** [comment]
Thanks for the help @sigprof I redid the layouts. Let me know if there is anything else I need to update.

File: keyboards/id87/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+		KC_GRV, KC_1, KC_2, KC_3, KC_4, KC_5, KC_6, KC_7, KC_8, 
+		KC_TAB, KC_Q, KC_W, KC_E, KC_R, KC_T, KC_Y, KC_U, KC_I, 
+		KC_CAPS, KC_A, KC_S, KC_D, KC_F, KC_G, KC_H, KC_J, KC_K, 
+		KC_LSFT, KC_Z, KC_X, KC_C, KC_V, KC_B, KC_N, KC_M, 
+		KC_LCTL, KC_LGUI, KC_LALT, KC_RCTL, MO(1), KC_SPC, KC_RGUI, KC_RALT, 
+		KC_PAUS, KC_SLCK, KC_F12, KC_F11, KC_F10, KC_PSCR, KC_F9, KC_F8, 
+		KC_0, KC_PGUP, KC_HOME, KC_BSPC, KC_EQL, KC_INS, KC_MINS, KC_9, 
+		KC_PGDN, KC_END, KC_BSLS, KC_RBRC, KC_LBRC, KC_DEL, KC_P, KC_O, 
+		KC_UP, KC_ENT, KC_QUOT, KC_SCLN, KC_L, 
+		KC_RGHT, KC_DOWN, KC_RSFT, KC_SLSH, KC_LEFT, KC_DOT, KC_COMM
```
[the remark this answers is not in the document]
Comment (mehoron) [reply keyboards/id87/keymaps/default/keymap.c]: Updated to use LT(1, KC_APP) instead. Verified it worked correctly with a few online key-testers.  (Still doesn't work in VIA).

File: keyboards/id87/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+                _______, _______, _______, _______, _______, _______, _______, _______, _______,
+                _______, _______, _______, _______, _______, _______, _______, _______, _______,
+                _______, _______, _______, _______, _______, _______, _______, _______,
+                _______, _______, _______, _______, _______, _______, _______, _______,
+                _______, _______, _______, _______, _______, _______, _______, _______,
+                _______, _______, _______, _______, _______, _______, _______, _______,
+                _______, _______, _______, _______, _______, _______, _______, _______,
+                _______, _______, _______, _______, _______,
+                _______, _______, _______, _______, _______, _______, _______
+                ),
```
[the remark this answers is not in the document]
Comment (mehoron) [reply keyboards/id87/keymaps/default/keymap.c]: Removed the empty layers from the default keymap.

File: keyboards/id87/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                _______, _______, _______, _______, _______, _______, _______, _______, _______,
+                _______, BL_TOGG, RGB_TOG, _______, _______, _______, _______, _______, _______,
+                _______, _______, _______, _______, _______, _______, _______, _______, _______,
+                _______, _______, _______, _______, _______, _______, NK_TOGG, _______,
+                _______, _______, _______, _______, _______, _______, _______, _______,
+                _______, _______, _______, _______, _______, _______, _______, _______,
+                _______, _______, _______, _______, _______, _______, _______, _______,
+                _______, _______, _______, _______, _______, _______, _______, _______,
+                BL_INC, _______, _______, _______, _______,
+                _______, BL_DEC, _______, _______, _______, _______, _______
```
[the remark this answers is not in the document]
Comment (mehoron) [reply keyboards/id87/keymaps/default/keymap.c]: Added default RGB commands back. Added additional commands and moved around the default

**Review by drashna — APPROVED** [review]

File: keyboards/id87/info.json [hunk 7]
```diff
@@ -0,0 +1,102 @@
+{
+    "keyboard_name": "ID87", 
+    "url": "https://www.idobao.net/products/idobao-id87-80-hot-swappable-mechanical-keyboard-kit", 
+    "maintainer": "qmk", 
+    "width": 18.25, 
+    "height": 6.5, 
+    "layouts": {
+        "LAYOUT_tkl_ansi": {
+            "key_count": 87,
```
Comment (fauxpark) [inline keyboards/id87/info.json:9]:
```suggestion
```

File: keyboards/id87/readme.md [hunk 8]
```diff
@@ -0,0 +1,19 @@
+# IDOBAO ID87
+
+![ID87](https://cdn.shopify.com/s/files/1/0382/1348/4675/products/6601EE887B14D36DE54CA137F0E383AE_1024x1024.jpg)
+
+*A TKL hotswap board from IDOBAO*
```
Comment (fauxpark) [inline keyboards/id87/readme.md:5]:
```suggestion
A TKL hotswap board from IDOBAO.
```

File: keyboards/id87/readme.md [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+* Keyboard Maintainer: Tybera
+* Hardware Supported: IDOBAO ID87, atmega32u4
+* Hardware Availability: [Idobao](https://www.idobao.net/products/idobao-id87-80-hot-swappable-mechanical-keyboard-kit)
+
+Make example for this keyboard (after setting up your build environment):
+
+    make id87:default
+
+Flash example for this keyboard:
+:
```
Comment (fauxpark) [inline keyboards/id87/readme.md:16]:
```suggestion
Flashing example for this keyboard:

```

File: keyboards/id87/rules.mk [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+CONSOLE_ENABLE = no			# Console for debug(+400)
+COMMAND_ENABLE = no			# Commands for debug and configuration
+SLEEP_LED_ENABLE = no		# Breathing sleep LED during USB suspend
+NKRO_ENABLE = yes			# USB Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE = yes      # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes		# Enable keyboard RGB underglow
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+LAYOUTS = tkl_ansi
```
Comment (fauxpark) [inline keyboards/id87/rules.mk:22]:
```suggestion
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = yes      # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output

LAYOUTS = tkl_ansi
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/id87/info.json [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+                {"label":"Win", "x":11.25, "y":5.5, "w":1.25},
+                {"label":"LT(1, KC_APP)", "x":12.5, "y":5.5, "w":1.25},
+                {"label":"Ctrl", "x":13.75, "y":5.5, "w":1.25},
+                {"label":"Left", "x":15.25, "y":5.5},
+                {"label":"Down", "x":16.25, "y":5.5},
+                {"label":"Right", "x":17.25, "y":5.5}
+            ]
+        }
+    }
+    ,"meta": "https://noroadsleft.github.io/kbf_qmk_converter/"
```
Comment (noroadsleft) [inline keyboards/id87/info.json:100]: I don't need you to shill for me. :)

```suggestion
```

**Review by noroadsleft — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Pull request merged**